src/App.vue

```diff
@@ -0,0 +1,92 @@
+<template>
+  <div class="container">
+    <div class="row">
+      <div class="col-4">
+        <MemoList @memoSelected="memoSelected" :memos="memos"></MemoList>
+        <AddButton @openNewMemo="openNewMemo"></AddButton>
+      </div>
+      <div class="col-8">
+        <EditForm v-if="editing || adding" @saveMemo="saveMemo" @delMemo="delMemo" :memo="this.editingMemo"></EditForm>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script>
+import MemoList from './components/MemoList.vue'
+import EditForm from './components/EditForm.vue'
+import AddButton from './components/AddButton.vue'
+
+const STORAGE_KEY = 'fjord-vuecli-memo'
+const memoStorage = {
+  fetch() {
+    return JSON.parse(localStorage.getItem(STORAGE_KEY) || '[]')
+  },
+  save(memos) {
+    localStorage.setItem(STORAGE_KEY, JSON.stringify(memos))
+  }
+};
+
+export default {
+  data(){
+    return {
+      memos: [],
+      editingMemo: [],
+      editing: false,
+      adding: false
+    }
+  },
+  mounted() {
+    this.memos = memoStorage.fetch()
+  },
+  components: {
+    MemoList,
+    EditForm,
+    AddButton
+  },
+  methods: {
+    memoSelected(memo){
+      this.editing = true
+      this.editingMemo = memo
+    },
+    openNewMemo() {
+      this.editingMemo = []
+      this.adding = true
+    },
+    saveMemo(editMemo){
+      if (!this.editingMemo.id) {
+        this.memos.push({
+          id: Date.now() + Math.floor(Math.random() * 100),
+          body: editMemo.body
+        })
+      }
+      memoStorage.save(this.memos)
+      this.editing = false
+      this.adding = false
```

Comment:
コメントありがとうございます。ご指摘頂いた

> 更新の処理自体が saveMemo の中に登場しないのはちょっと分かりにくい

という点について、問題点が明確に理解できずにいるため、すみませんがもう少し教えて下さい。
（自分の書いたコードに`saveMemo`というメソッドが`EditForm.vue`と`App.vue`の両方にある事自体でわかりにくくしてしまっているのですが）ご指摘いただいた内容は「メモの保存や追加（localStorage.setItem）（削除も？）はEditForm.vue内で行うべき」ということでしょうか。

また、２と３のemitによるイベントの流れというのは、MemoListやAddButtonでメモ（や新規）を選択されたら、emitによりEditFormでイベントを発火させる（それによってwatchが不要になる？）という理解で合っていますでしょうか。
$emitは子から親コンポーネントに値を渡すと認識していたので、↑はその逆もあり？という点で多少混乱気味な感じです。修正の実装に着手する前に少し整理する必要がありそうなので、お手数ですがコメントよろしくお願いします。

Reply:
これは説明するのがかなり難しいのですが、メモの編集時には以下のようなことが起こっているんですよね。`App.vue` の `this.editingMemo` と `EditForm.vue` の `this.memo` および `this.editMemo` がすべて同じものを指しているので、更新のときの処理がうまく動く、けどそれは設計としてちょっと分かりづらいかも、というのが簡単な説明になります。

このフィードバックで伝えたいことは 2 つあって、

* 子コンポーネントで親コンポーネントのデータを書き換える形になってしまっているので、そうしないようにしよう
* 上の副作用に依存して更新処理が行われているので、副作用に依存しない形で更新処理を行えるようにしよう

ということでした。1 つめのフィードバックは、具体的には `EditForm.vue` の `refreshMemo()` メソッドの中で `this.editMemo = this.memo` を実行しているところの話になります。JavaScript ではオブジェクトを別の変数に代入してもオブジェクトコピーされるわけではなく、同じオブジェクトを指すようになることに加えて、Vue.js の props でもそれと同じような挙動になる、ということがこの問題を引き起こしています。

> When objects and arrays are passed as props, while the child component cannot mutate the prop binding, it will be able to mutate the object or array's nested properties. This is because in JavaScript objects and arrays are passed by reference, and it is unreasonably expensive for Vue to prevent such mutations.

https://vuejs.org/guide/components/props.html#one-way-data-flow

とはいえ、どこから手を付けるといいかはなかなか難しいところだと思うので、まずは `EditForm.vue` の `refreshMemo()` メソッドで `this.editMemo = this.memo` と実行しているところを `this.editMemo = { ...this.memo }` (オブジェクトをコピーするイディオムです) のようにしてみるところから始めると良いかもしれません。そうすると、1 つ前のコメントの話 (https://github.com/NorifumiOgawa/vuecli-memo/pull/1#discussion_r803714819) に自然とつながってくると思います。

* * *

## 初期状態

```
localStorage:      [{ id: 123456789, body: 'Hello World' }, ...]
App.memos:         [{ id: 123456789, body: 'Hello World' }, ...]
App.editingMemo:   null
EditForm.editMemo: ''
```

## 編集するメモを選択する

1. `MemoList.memoSelected(memo)`
2. `MemoList.$emit('memoSelected', memo)`
3. `App.memoSelected(memo)`

```
localStorage:      [{ id: 123456789, body: 'Hello World' }, ...]
App.memos:         [{ id: 123456789, body: 'Hello World' }, ...]
App.editingMemo:   { id: 123456789, body: 'Hello World' }
EditForm.editMemo: ''
```

4. (`EditForm` がレンダリングされる)
5. `MemoList.mounted()`
6. `MemoList.refreshMemo()`

```
localStorage:      [{ id: 123456789, body: 'Hello World' }, ...]
App.memos:         [{ id: 123456789, body: 'Hello World' }, ...]
App.editingMemo:   { id: 123456789, body: 'Hello World' }
EditForm.editMemo: { id: 123456789, body: 'Hello World' }
```

## メモを編集して保存する

1. `EditForm.saveMemo()`

```
localStorage:      [{ id: 123456789, body: 'Hello World' }, ...]
App.memos:         [{ id: 123456789, body: 'こんにちは世界' }, ...]
App.editingMemo:   { id: 123456789, body: 'こんにちは世界' }
EditForm.editMemo: { id: 123456789, body: 'こんにちは世界' }
```

2. `EditForm.$emit('saveMemo', this.editMemo)`

```
localStorage:      [{ id: 123456789, body: 'Hello World' }, ...]
App.memos:         [{ id: 123456789, body: 'こんにちは世界' }, ...]
App.editingMemo:   { id: 123456789, body: 'こんにちは世界' }
EditForm.editMemo: { id: 123456789, body: 'こんにちは世界' }
```

3. `App.saveMemo(editMemo)`

```
localStorage:      [{ id: 123456789, body: 'こんにちは世界' }, ...]
App.memos:         [{ id: 123456789, body: 'こんにちは世界' }, ...]
App.editingMemo:   { id: 123456789, body: 'こんにちは世界' }
EditForm.editMemo: { id: 123456789, body: 'こんにちは世界' }
```

Reply:
コメントありがとうございました。
ご指摘いただいた、

> まずは `EditForm.vue` の `refreshMemo()` メソッドで `this.editMemo = this.memo` と実行しているところを `this.editMemo = { ...this.memo }` (オブジェクトをコピーするためのイディオムです) のようにしてみる

という部分から着手してみました。
`App.vue`からpropsで受け取った`memo`を`this.editMemo`にコピーし、テキストエリアのv-modelには、その`this.editMemo.body`を設定してみました。
それによって、`EditForm.vue`内の`memoBody`は全て不要になったので削除しました。

また、`EditForm.vue`のメソッド`saveMemo`を
- localstorageへのデータ保存はしていないのと、
- `EditForm.vue`でイベントを発火させて`editMemo`の内容を`App.vue`に渡すという動き
から、`emitMemo`という名称に変更しました。

この修正を加えた後も一応要件通りの動作はしているのですが、指摘していただいた以下２点の部分について理解（と対応）ができている感じが無いのですが、現状、以下のように捉えています。

> 子コンポーネントで親コンポーネントのデータを書き換える形になってしまっているので、そうしないようにしよう

ここについては、`this.editMemo = { ...this.memo }`とすることで、同じオブジェクトを指すのではなく、オブジェクトをコピーしているので、親コンポーネントのデータ（=this.memo?）を触ることは無くなるという理解で良い？

> 副作用に依存して更新処理が行われている

今回`副作用`になっていたのは、propsで受け取った`memo`を`this.editMemo = this.memo`としていたことで、`EditForm.vue`内でも同じオブジェクト（＝親コンポーネントのデータ？）を操作・参照していることを指していたという理解で良い？

もし対応が間違っていたらご指摘お願いします。何度もすみません！

Reply:
> > 子コンポーネントで親コンポーネントのデータを書き換える形になってしまっているので、そうしないようにしよう
> 
> ここについては、`this.editMemo = { ...this.memo }`とすることで、同じオブジェクトを指すのではなく、オブジェクトをコピーしているので、親コンポーネントのデータ（=this.memo?）を触ることは無くなるという理解で良い？

はい、あっています！

> > 副作用に依存して更新処理が行われている
> 
> 今回`副作用`になっていたのは、propsで受け取った`memo`を`this.editMemo = this.memo`としていたことで、`EditForm.vue`内でも同じオブジェクト（＝親コンポーネントのデータ？）を操作・参照していることを指していたという理解で良い？

副作用になっていたのは `saveMemo()` の `this.editMemo.body = this.memoBody` の部分です！ `App.vue` の `this.editingMemo` と `EditForm.vue` の `this.editMemo` が同じオブジェクトを指していたので、`EditForm.vue` の `this.editMemo` の状態を変更する (= `this.editMemo.body` を書き換える) と `App.vue` の `this.editingMemo` の状態も変更される (= `this.editingMemo.body` が書き換えられる) 、というのが副作用でした。より正確には「〜を変更すると〜も変更される」のではなく、どちらも同じものを指しているのでそれが単に書き換えられただけ、ということになります。

変更前のコードのデータや変数の様子を図解してみたので、よかったら参考にしてみてください！

![IMG_2050](https://user-images.githubusercontent.com/7645585/158042866-4143901f-25a7-4fc2-82f7-78fce35856ff.jpeg)